[FIX] mail: make file upload consistent between guest and portal

seb-odoo opened the pull request.

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/82912).

mart-e commented on addons/mail/models/mail_message.py:
```diff
@@ -911,9 +911,9 @@ def _message_fetch(self, domain, max_id=None, min_id=None, limit=30):
         if min_id:
             domain = expression.AND([domain, [('id', '>', min_id)]])
-        return self.search(domain, limit=limit).message_format()
+        return self.search(domain, limit=limit).message_format(format_attachments_token='if-author' if self.env.user.share else False)
```
Nitpicking: the new parameter has 2 possible values: `'if-author'` and `False`. Why not a simple boolean?

seb-odoo replied:
Because `True` looks like it should always be set, but we only want to set it if the current user is author.
It could be `True` if the param was named `format_attachments_token_if_author` but that looked a bit too long.
I don't really mind either way if there is a strong consensus for one or the other.

xmo-odoo replied:
Given this is a Python -> Python communication, maybe an `Enum` would be a good option? Or a signaling `object()` checked by identity? That's a pretty common pattern.

seb-odoo replied:
I've never used those before, so just `IF_AUTHOR = object()` at the top of the file to avoid using hardcoded string?

mart-e commented on addons/mail/controllers/discuss.py:
```diff
@@ -196,6 +196,8 @@ def discuss_starred_messages(self, max_id=None, min_id=None, limit=30, **kwargs)
     @http.route('/mail/message/post', methods=['POST'], type='json', auth='public')
     def mail_message_post(self, thread_model, thread_id, post_data, **kwargs):
+        attachment_ids = post_data.get('attachment_ids', [])
+        request.env['mail.thread']._message_post_check_attachments(attachment_ids=attachment_ids, attachment_tokens=post_data.get('attachment_tokens', [None] * len(attachment_ids)))
```
why the `[None] * len(attachment_ids)` ? The check is going to fail anyway if there is not tokens.

seb-odoo replied:
Tokens are only checked if the current user does not naturally have access to the document, which could be the case for internal users. (cf. the methods that have been moved from portal)

This ensures the most possible compatibility (running the old JS on the new server, or the old server with the new JS). In most situations it will just keep working like before, and only for guest both need to be fully updated.

xmo-odoo commented on addons/mail/models/ir_attachment.py:
```diff
@@ -44,7 +44,7 @@ def _delete_and_notify(self):
         self.unlink()
 
-    def _attachment_format(self, commands=False):
+    def _attachment_format(self, commands=False, format_token=False):
```
The parameter name is quite confusing, it doesn't actually format anything. Maybe `include_token` would be clearer?

xmo-odoo commented on addons/mail/models/mail_thread.py:
```diff
@@ -1646,6 +1646,17 @@ def _message_partner_info_from_emails(self, emails, link_mail=False):
+        should not be called from a sudo env."""
+        if len(attachment_tokens) != len(attachment_ids):
+            raise UserError(_("An access token must be provided for each attachment."))
+        for (attachment_id, access_token) in zip(attachment_ids, attachment_tokens):
```
The parens are unnecessary.

xmo-odoo commented on addons/mail/models/ir_http.py:
```diff
@@ -29,3 +31,18 @@ def session_info(self):
             })
         return result
+
+    def _document_check_access(self, model_name, document_id, access_token=None):
```
This is apparently moved from portal, shouldn't it also be *re*moved from there?

seb-odoo replied:
I believe I did remove it from portal, at least the code itself, not the method.

However I didn't adapt all existing occurrences but simply did a forward for now: https://github.com/odoo/odoo/pull/82912/files#diff-2acd0e6f67a1190985466e669bc815626745d14981784d828e404f8855af186eR366

Do you suggest that I go over the 26 existing occurrences across various modules to make the use the new method instead and get rid of the old one completely? That looked a bit overkill in stable, but if you think it's better I can do it.

tde-banana-odoo replied:
Not sure what 'model_name' is, but as anyway you expect an access_token field and it seems called only on ir attachment, why not moving it as a model method on attachment model ? Not sure why we need this kind of generic methods on ir_http, nothing to do with routing code unless I misunderstood something ?

seb-odoo replied:
This is a move of the portal method, it's called with all kind of `model_name` as parameter in the various portal controllers.
It's not related to attachments specifically, it's actually closely related to the portal mixin (but in reality, just the presence of an `access_token` field on the target model is enough).

Except now we also need it in `mail` (in this case for attachments), but it's not in the dependency for portal (and attachment doesn't have the portal mixin anyway).

I could move it into models.py override instead of http file?

seb-odoo replied:
I put it originally in http file was because this method is only called in the context of a request for public user/with token before granting sudo access.

tde-banana-odoo replied:
Maybe it should be on models.py then, without model_name. Not sure why people try to put model methods as generic helpers. We have models, let us use them.

xmo-odoo commented on addons/mail/models/mail_thread.py:
```diff
@@ -1646,6 +1646,17 @@ def _message_partner_info_from_emails(self, emails, link_mail=False):
     # MESSAGE POST API
     # ------------------------------------------------------
 
+    def _message_post_check_attachments(self, attachment_ids, attachment_tokens):
```
Given `mail.thread` is a mixin, I don't know that this API is sensible when it can (and should?) not be overridden on a per-model basis.

tde-banana-odoo replied:
It has effectively nothing to do with standard mail.thread code ... just internals related to attachments. I don't think we really want to enable model-based management, even with a static management of attachments it is already a complete mess ^^ .

xmo-odoo replied:
I think either a free function (to import normally) or on mail.message would make sense.

Alternatively, a sibling function of `_document_check_access`: in essence it's just a way to easily mass-call that, so either put them next to one another, or just have `_document_check_access` work on multiple records and tokens directly.

seb-odoo replied:
Ok that looks like a good idea to only have one function, I guess I'll just copy the code from portal and adapt it to work specifically in batch for attachments, rather than trying to re-use it maybe incorrectly.

tde-banana-odoo replied:
Does not seem message model has any link with this, why would we put that on message model ? But if this can be done generic (on BaseModel ?), better indeed.

seb-odoo commented:
Above force-push is rebase, I'll check the comments now.

tde-banana-odoo commented on addons/mail/models/mail_message.py:
```diff
@@ -15,6 +15,8 @@
 _logger = logging.getLogger(__name__)
 _image_dataurl = re.compile(r'(data:image/[a-z]+?);base64,([a-z0-9+/\n]{3,}=*)\n*([\'"])(?: data-filename="([^"]*)")?', re.I)
 
+IF_AUTHOR = object()
```
What is that ?

xmo-odoo replied:
See https://github.com/odoo/odoo/pull/82912#discussion_r786971594 I suggested an enum (like the `odoo.fields.Command`) but seb preferred a placeholder object. My goal was mostly to avoid the weird magic string thing since for clarity they preferred avoiding a boolen

seb-odoo replied:
I feel the extra syntax needed for Enum is overkill for this very specific case, it's not really worth having extra stuff at the top of the file.

Do we absolutely want to avoid the hardcoded string? I feel like it was the best compromise in this case.

tde-banana-odoo reviewed (commented):
Pingpong

seb-odoo commented:
Thank you for the reviews.

PR adapted to comments:
- moved check method to attachment and made it specific
- reverted most useless diff in portal and http files
- added test
- changed formatter param to be boolean (I'm still not happy with this part, I'll see what I can do, maybe to get rid of the param altogether)

seb-odoo commented:
PR updated to get rid of the annoying extra parameters

beledouxdenis commented:
Is there something needed to proceed to the merge of this ?

If it's blocked by fear of regressions in stable, to me, this could totally land in master (and optionally later merged in stable once the patch is confirmed working in master, though I don't think it's even needed).

tde-banana-odoo commented:
I suppose beginning by writing a commit message would help reviewing.
